// @azure/cosmos/dist-esm/queryIterator.js
import * as tslib_1 from "tslib";
import { getPathFromLink, ResourceType, StatusCodes, SubStatusCodes } from "./common";
import { DefaultQueryExecutionContext, getInitialHeader, mergeHeaders, PipelinedQueryExecutionContext } from "./queryExecutionContext";
import { FeedResponse } from "./request/FeedResponse";
/**
 * Represents a QueryIterator Object, an implementation of feed or query response that enables
 * traversal and iterating over the response
 * in the Azure Cosmos DB database service.
 */
export class QueryIterator {
    /**
     * @hidden
     */
    constructor(clientContext, query, options, fetchFunctions, resourceLink, resourceType) {
        this.clientContext = clientContext;
        this.query = query;
        this.options = options;
        this.fetchFunctions = fetchFunctions;
        this.resourceLink = resourceLink;
        this.resourceType = resourceType;
        this.query = query;
        this.fetchFunctions = fetchFunctions;
        this.options = options;
        this.resourceLink = resourceLink;
        this.fetchAllLastResHeaders = getInitialHeader();
        this.reset();
    }
    /**
     * Gets an async iterator that will yield results until completion.
     *
     * NOTE: AsyncIterators are a very new feature and you might need to
     * use polyfils/etc. in order to use them in your code.
     *
     * If you're using TypeScript, you can use the following polyfill as long
     * as you target ES6 or higher and are running on Node 6 or higher.
     *
     * ```typescript
     * if (!Symbol || !Symbol.asyncIterator) {
     *   (Symbol as any).asyncIterator = Symbol.for("Symbol.asyncIterator");
     * }
     * ```
     *
     * @example Iterate over all databases
     * ```typescript
     * for await(const {result: db} in client.databases.readAll().getAsyncIterator()) {
     *   console.log(`Got ${db.id} from AsyncIterator`);
     * }
     * ```
     */
    getAsyncIterator() {
        return tslib_1.__asyncGenerator(this, arguments, function* getAsyncIterator_1() {
            this.reset();
            this.queryPlanPromise = this.fetchQueryPlan();
            while (this.queryExecutionContext.hasMoreResults()) {
                let response;
                try {
                    response = yield tslib_1.__await(this.queryExecutionContext.fetchMore());
                }
                catch (error) {
                    if (this.needsQueryPlan(error)) {
                        yield tslib_1.__await(this.createPipelinedExecutionContext());
                        response = yield tslib_1.__await(this.queryExecutionContext.fetchMore());
                    }
                    else {
                        throw error;
                    }
                }
                const feedResponse = new FeedResponse(response.result, response.headers, this.queryExecutionContext.hasMoreResults());
                if (response.result !== undefined) {
                    yield yield tslib_1.__await(feedResponse);
                }
            }
        });
    }
    /**
     * Determine if there are still remaining resources to processs based on the value of the continuation token or the\
     * elements remaining on the current batch in the QueryIterator.
     * @returns {Boolean} true if there is other elements to process in the QueryIterator.
     */
    hasMoreResults() {
        return this.queryExecutionContext.hasMoreResults();
    }
    /**
     * Fetch all pages for the query and return a single FeedResponse.
     */
    fetchAll() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.reset();
            this.fetchAllTempResources = [];
            return this.toArrayImplementation();
        });
    }
    /**
     * Retrieve the next batch from the feed.
     *
     * This may or may not fetch more pages from the backend depending on your settings
     * and the type of query. Aggregate queries will generally fetch all backend pages
     * before returning the first batch of responses.
     */
    fetchNext() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.queryPlanPromise = this.fetchQueryPlan();
            let response;
            try {
                response = yield this.queryExecutionContext.fetchMore();
            }
            catch (error) {
                if (this.needsQueryPlan(error)) {
                    yield this.createPipelinedExecutionContext();
                    response = yield this.queryExecutionContext.fetchMore();
                }
                else {
                    throw error;
                }
            }
            return new FeedResponse(response.result, response.headers, this.queryExecutionContext.hasMoreResults());
        });
    }
    /**
     * Reset the QueryIterator to the beginning and clear all the resources inside it
     */
    reset() {
        this.queryPlanPromise = undefined;
        this.queryExecutionContext = new DefaultQueryExecutionContext(this.options, this.fetchFunctions);
    }
    toArrayImplementation() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.queryPlanPromise = this.fetchQueryPlan();
            while (this.queryExecutionContext.hasMoreResults()) {
                let response;
                try {
                    response = yield this.queryExecutionContext.nextItem();
                }
                catch (error) {
                    if (this.needsQueryPlan(error)) {
                        yield this.createPipelinedExecutionContext();
                        response = yield this.queryExecutionContext.nextItem();
                    }
                    else {
                        throw error;
                    }
                }
                const { result, headers } = response;
                // concatenate the results and fetch more
                mergeHeaders(this.fetchAllLastResHeaders, headers);
                if (result !== undefined) {
                    this.fetchAllTempResources.push(result);
                }
            }
            return new FeedResponse(this.fetchAllTempResources, this.fetchAllLastResHeaders, this.queryExecutionContext.hasMoreResults());
        });
    }
    createPipelinedExecutionContext() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const queryPlanResponse = yield this.queryPlanPromise;
            // We always coerce queryPlanPromise to resolved. So if it errored, we need to manually inspect the resolved value
            if (queryPlanResponse instanceof Error) {
                throw queryPlanResponse;
            }
            const queryPlan = queryPlanResponse.result;
            const queryInfo = queryPlan.queryInfo;
            if (queryInfo.aggregates.length > 0 && queryInfo.hasSelectValue === false) {
                throw new Error("Aggregate queries must use the VALUE keyword");
            }
            this.queryExecutionContext = new PipelinedQueryExecutionContext(this.clientContext, this.resourceLink, this.query, this.options, queryPlan);
        });
    }
    fetchQueryPlan() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.queryPlanPromise && this.resourceType === ResourceType.item) {
                return this.clientContext
                    .getQueryPlan(getPathFromLink(this.resourceLink) + "/docs", ResourceType.item, this.resourceLink, this.query, this.options)
                    .catch((error) => error); // Without this catch, node reports an unhandled rejection. So we stash the promise as resolved even if it errored.
            }
            return this.queryPlanPromise;
        });
    }
    needsQueryPlan(error) {
        return (error.code === StatusCodes.BadRequest &&
            error.substatus &&
            error.substatus === SubStatusCodes.CrossPartitionQueryNotServable);
    }
}
