// @azure/cosmos/dist-esm/queryExecutionContext/EndpointComponent/UnorderedDistinctEndpointComponent.js
import * as tslib_1 from "tslib";
import { sha1 } from "crypto-hash";
import stableStringify from "fast-json-stable-stringify";
/** @hidden */
export class UnorderedDistinctEndpointComponent {
    constructor(executionContext) {
        this.executionContext = executionContext;
        this.hashedResults = new Set();
    }
    nextItem() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { headers, result } = yield this.executionContext.nextItem();
            if (result) {
                const stringifiedResult = stableStringify(result);
                const hashedResult = yield sha1(stringifiedResult);
                if (this.hashedResults.has(hashedResult)) {
                    return { result: undefined, headers };
                }
                this.hashedResults.add(hashedResult);
            }
            return { result, headers };
        });
    }
    current() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.executionContext.current();
        });
    }
    hasMoreResults() {
        return this.executionContext.hasMoreResults();
    }
}
