// @azure/cosmos/dist-esm/retry/endpointDiscoveryRetryPolicy.js
import * as tslib_1 from "tslib";
import { isReadRequest } from "../common/helper";
/**
 * This class implements the retry policy for endpoint discovery.
 * @hidden
 */
export class EndpointDiscoveryRetryPolicy {
    /**
     * @constructor EndpointDiscoveryRetryPolicy
     * @param {object} globalEndpointManager The GlobalEndpointManager instance.
     */
    constructor(globalEndpointManager, operationType) {
        this.globalEndpointManager = globalEndpointManager;
        this.operationType = operationType;
        this.maxRetryAttemptCount = EndpointDiscoveryRetryPolicy.maxRetryAttemptCount;
        this.currentRetryAttemptCount = 0;
        this.retryAfterInMilliseconds = EndpointDiscoveryRetryPolicy.retryAfterInMilliseconds;
    }
    /**
     * Determines whether the request should be retried or not.
     * @param {object} err - Error returned by the request.
     */
    shouldRetry(err, retryContext, locationEndpoint) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!err) {
                return false;
            }
            if (!retryContext || !locationEndpoint) {
                return false;
            }
            if (!this.globalEndpointManager.enableEndpointDiscovery) {
                return false;
            }
            if (this.currentRetryAttemptCount >= this.maxRetryAttemptCount) {
                return false;
            }
            this.currentRetryAttemptCount++;
            if (isReadRequest(this.operationType)) {
                this.globalEndpointManager.markCurrentLocationUnavailableForRead(locationEndpoint);
            }
            else {
                this.globalEndpointManager.markCurrentLocationUnavailableForWrite(locationEndpoint);
            }
            // Check location index increment
            // TODO: Tracing
            // console.log("Write region was changed, refreshing the regions list from database account
            // and will retry the request.");
            yield this.globalEndpointManager.refreshEndpointList();
            retryContext.retryCount = this.currentRetryAttemptCount;
            retryContext.clearSessionTokenNotAvailable = false;
            retryContext.retryRequestOnPreferredLocations = false;
            return true;
        });
    }
}
EndpointDiscoveryRetryPolicy.maxRetryAttemptCount = 120; // TODO: Constant?
EndpointDiscoveryRetryPolicy.retryAfterInMilliseconds = 1000;
