// @azure/cosmos/dist-esm/queryExecutionContext/EndpointComponent/OrderedDistinctEndpointComponent.js
import * as tslib_1 from "tslib";
import { sha1 } from "crypto-hash";
import stableStringify from "fast-json-stable-stringify";
/** @hidden */
export class OrderedDistinctEndpointComponent {
    constructor(executionContext) {
        this.executionContext = executionContext;
    }
    nextItem() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { headers, result } = yield this.executionContext.nextItem();
            if (result) {
                const stringifiedResult = stableStringify(result);
                const hashedResult = yield sha1(stringifiedResult);
                if (hashedResult === this.hashedLastResult) {
                    return { result: undefined, headers };
                }
                this.hashedLastResult = hashedResult;
            }
            return { result, headers };
        });
    }
    current() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.executionContext.current();
        });
    }
    hasMoreResults() {
        return this.executionContext.hasMoreResults();
    }
}
